library/Fission/User/DID/Types.hs

```diff
@@ -1,27 +1,117 @@
+> 0x7a 0xed01 ED25519_PUBLIC_KEY_BYTES
+> [...]
+> Expressing a cryptographic identifier that is a base58btc encoded RSA SPKI-based
+>   public key fingerprint using SHA2-256/256 would look like this:
+> 0x7a 0x5d 0x12 0x20 HASH_BYTES
+https://github.com/w3c-ccg/lds-ed25519-2018/issues/3
+
+> So the procedure to express other types of keys is:
+>
+>    1. Find the multicodec representation for your key type.
+>         For example, the registration for secp256k1 keys is pending
+>         over at multiformats/multicodec#160.
+>    2. Encode the raw public key bytes using that multicodec value.
+>    3. Pick a character encoding (I suggest base58-btc),
+>         and convert it to a character string using Multibase.
+>    4. Prepend did:key: to that string, and there you have it!
+https://github.com/w3c-ccg/did-method-key/issues/3#issuecomment-594190524
+
+NOTE: The multihash is in a middle-endian format -- SB base-7 with continuation flags
+More here: https://github.com/multiformats/unsigned-varint
+
+==== __Examples__
+
+>>> decode' "\"did:key:zBR4m3DNZHT1G8Nb2RHzgKK7TrWxEmJjZskgvFeJwYJ6kpzy1PVDvn3jR2vaAWExNdtKT7KzBoAdy8GHeGd8jpiAUDgbRRnMy\"" :: Maybe DID
+Just (DID {publicKey = Public {publicKey = "AAAAC3NzaC1lZDI1NTE5AAAAIIPnL+R9+OrIm26I1MSOnu4ofAtJ5PjmfiO9ukShjoST"}, algorithm = Ed25519, method = Key})
+
+>>> decode' "\"did:key:z1ArJG77479Lcdo6AYX1efdvVo2dhSeAwPpazVKhFb8Sg4w35AKjx4sdfLy36LnuBq8b3JGt4SJCsnT3mmnujxxehgTiNcXw495k5dafYE9Zt77T8xCWdDkrLLvbGBzV17vNunkgkZKaA5WqwhV4QmFdFCCvEHszNe62SxFCR6g2YqAimX1DSWwVQFc12iuRYrXj8nWeMv2iQG3ZMCvKo2XsW7SXR1Kai29HZdwFQQneEdoEbTWHtFRUh5MpZuX95WNkaZy4JWHjQUNGgNRM6gGrgnmF5ibbXXoTNGhwCH2Qt9DR9khd6t38ufNUdcffapWk5cuGo5pfM1Zye9vapGMpLxQgRUnSyXiz41tSbMHzveTYggiCShVZJ27VLigREDMBqbU8bUNXZBL5L2yTaQieUhPbHsuM1FUD4AGfGQwAmdUstsMC4tjLoYjRBe6vJnwMMxB2ks89M2Yu2aYPf5WgkRGAX9EeXmmiq72A2dPThP4XGaP8pmWw7658BUTuA\"" :: Maybe DID
+Just (DID {publicKey = Public {publicKey = "AAAAB3NzaC1yc2EAAAADAQABAAABAQDkrRwcO9XZOWdwcK9CUQbzD3NMGlmkoRWu/BS5b/C9lm7PIyjBIhshnd6Y29upBKra7dJ7b1qOJDRQS5uvu93OZi/6pGXcqlYHS9WWJtpEQM+VXeJ2PcnKl5ok2mWgeOEqjHRorT+2dVlISjvOk4dRTJR2sB3el8ynQ1W7LuiEio22352O0DYV89DMhMPVVoSvXVBbsvuJv4VJ4e2XYlilsYyF/6zba4rvEP37MJBExNUqlWUbmIAzFbSoJSdickzHJtLCaBu8Eapu/bu90ecNiFIEaXDSvjD+wVqNwqaarWDor248BULN0u3mVTxHh185k8kBAK6ITBnDMJzjsk11"}, algorithm = RSA2048, method = Key})
+
+-}
+data DID = DID
+  { publicKey :: !Key.Public
+  , algorithm :: !Key.Algorithm
+  , method    :: !Method
+  } deriving (Show, Eq)
+
+instance Arbitrary DID where
+  arbitrary = do
+    publicKey <- arbitrary
+    algorithm <- arbitrary
+    method    <- arbitrary
+
+    return DID {..}
+
+instance ToJSON DID where
+  toJSON (DID (Key.Public pk) algo method) =
+    String (header <> UTF8.toBase58Text multicodec64)
+    where
+      header :: Text
+      header = "did:" <> textDisplay method <> ":" <> "z"
+
+      multicodec64 :: ByteString
+      multicodec64 = BS.pack magicBytes <> encodeUtf8 pk
```

Comment:
These aren't base64 encoded so we probably want to change the name of this variable

I'm not sure what `BS.pack` actually encodes to, but it's what the base58 package does under the hood so that seems right

`encodeUtf8 pk` encodes to UTF-8 so we're sending `base58(utf8(pk))`. where it looks like pk is already base58 encoded text. so it's actually `base58(utf8(base58(pk)))`. Wouldn't it be better to just send `base58(pk)`? I know we talked about something similar on our call, but I can't quite remember the details, so I might be forgetting the reason we're doing this

Reply:
@dholms 

> These aren't base64 encoded so we probably want to change the name of this variable

I'm very confused then 😛  Here's my thinking:

* `magicBytes` are binary
  * Entered in hex (base16)
  * Rendered down to `[Word8]`s (lists of base8)

* The type of is `pk :: Text`, which is UTF8, which can technically be in any base
  * What you do with it from there is up to you

My understanding is that you can't just append arbitrary base58 together. I've been double checking my work against a base58 converter.

```
| Base16       | Base64 | Base58   |
|--------------|--------|----------|
| 616263       | abc    | ZiCa     |
|       646566 |    def | aisf     |
| 616263646566 | abcdef | qVgfxYy3 | <- Lines 1 & 2 of base64 appended
| 41aff07b7296 | Aﾯ￰{rﾖ  | ZiCaaisf | <- Lines 1 & 2 of base58 appended
```

Since the actual multicodec flags are hex, as I assume is the key, my understanding is that we can't just append text together and call it a day. 

We need to _first render the entire string down to binary_, and then encode in base58. Here's the format:

```
<MULTIBASE(base58-btc, MULTICODEC(public-key-type, raw-public-key-bytes))>
                                                   ^^^^^^^^^^^^^^^^^^^^
                                                    MAKE THE KEY BINARY
```

So in pseudo-code:

```haskell
base58key
  |> toBinary
  |> append(magicNumberBinary)
  |> encodeBase58
  |> append("z")
```

But I can also be TOTALLY WRONG. This stuff is confusing.

Reply:
yeah sorry if I was unclear 

3 points:

1) `multicode64`: I was mostly remarking on the variable name here. None of these values are base64, `BS.pack magicBytes` is raw binary (packed as a bytestring) and `encodeUtf8 pk` is a utf-8 encoded base58 key. a better variable name might just be `multicodec`

2) `BS.pack magicBytes` is raw binary packed into a bytestring. This seems right :+1: 

3) `encodeUtf8 pk` does not go to the raw public key bytes. `pk` is already a base-58 representation of the key. So `encodeUtf8 pk = utf8(base58(raw-public-key-bytes))`

In other words, the pseudo code is right, but the key is not being properly translated to binary
```haskell
base58key
  |> toBinary -- we're not actually going to the raw binary of the key
  |> append(magicNumberBinary)
  |> encodeBase58
  |> append("z")
```

unless I'm misunderstanding the code, right now the `toBinary` step is `encodeUtf8 pk`

it should be something like:
```haskell
toBinary :: Public -> ByteString
toBinary = BS58.BTC.toBytes . BS58.BTC.fromText . publicKey
```

also agreed on base-58 concatenation not working :+1: I don't think arbitrary concatenation works unless the base is an exponent of 2

Reply:
> pk is already a base-58 representation of the key.

Ah, yeah, good point! This is already in base58 because it's coming off the wire with the right _characters_ but in base64. So yes — reinterpret as base58 with `fromText`!

```
toBinary :: Public -> ByteString
toBinary = BS58.BTC.toBytes . BS58.BTC.fromText . publicKey
```

Right, exactly like what I have in the deserializer :P


Thanks!

Reply:
Oh wait, actually no.

The `pk` in `ToJSON` is a `Text` value that we previously stored. _It's stored in base64._

Reply:
I thought `pk` was a Base-58 string from looking at its `Arbitrary` instance. Even tho it's encoded base2 by `Text`. So like base2 encoded base-58 string (this stuff is so confusing to talk about 😛)

> multicodecW8 is a better name
yeah I like that

Reply:
Oh, yeah, confusing for sure 💯 `Arbitrary` gives us our consistent _internal_ representation, which is base2 `Text`

Reply:
Oh, you mean the `Arbitrary Key.Public` instance. Yeah... that is literally just random  bytes. The base58 stuff there is just me being confused about base vs text encoding 😅

Reply:
But it's a base2 `Text` representation of a base58 string right?
```haskell
instance Arbitrary Public where
  arbitrary = do
    nonempty :: NonEmpty Word8 <- arbitrary
    let pk = BS58.BTC.toText . BS58.BTC.fromBinary $ NonEmpty.toList nonempty
    return (Public pk)
```

Reply:
Comment at the same time! 👏

It is yeah. But we don't care about that for our internal representation. I've changed it to this just now:

```
instance Arbitrary Public where
  arbitrary = Public <$> arbitrary
```

Reply:
ah okay cool 👍 so what's the expected encoding of a key then? Base64? (that's what I use in `keystore-idb` for instance)

Reply:
The key itself is binary — so base64 is a pretty natural way to turn that into text 👍

Inside the Haskell application, it's raw binary again. When we read in a key, the default decoder will treat the incoming text as base64. When sending us request authed by a UCAN, it'll need to include this `did:key:z<BASE58>` stuff, but you can store the key in whatever format on the client 👍